# Part 1
### Define and explain the graphs being used to model the android applications (from the Hindroid paper). Explain any computationally relevant portions of the process of writing (text processing) code to extract the nodes and edges of the graphs.
The graphs in used to model Android applications are:
* app-to-api: maps apps to all the different APIs that they use
* api-to-api: maps APIs to APIs based on if they coexist in the same code block
* api-to-api: maps APIs to APIs based on if they are in the same package. To save space this is done as a mapping of API<-> packages. In order to get API<->API we can simply do AA<sup>T</sup>
* api-to-api: maps APIs to APIs based on if they are invoked in the same way. To save space this is done as a mapping of API<->invoke method. To get API<->API we can simply do AA<sup>T</sup>

In order to text process, we need to extract call methods, remember in the methods what APIs we have called, and have some way of associating each method call with a respective node in our graph. Then, we can simply get those nodes out to create edges between them.

In [3]:
from data_pipeline import download_and_process_apks

download_and_process_apks(50)

ignoring because size is 71.600000
ignoring because size is 626.600000
ignoring 2048 because exists
ignoring 2048 because exists
4636 downloaded
ignoring because size is 84.900000
five-craft-nights downloaded
modern-sniper downloaded
ignoring guns because exists
dual downloaded
ignoring flappy-crush because exists
batman-the-flash-hero-run downloaded
gun-shoot-war downloaded
ignoring because size is 69.600000
sonic-jump-fever downloaded
ignoring because size is 99.500000
ignoring because size is 89.100000
ignoring because size is 75.700000
prop-hunt-portable downloaded
deathrun-portable downloaded
ignoring because size is 61.700000
ignoring because size is 53.600000
mountain-sniper-shooting-3d-fps downloaded
ignoring ninja-revenge because exists
zombie-frontier downloaded
ignoring birds-vs-zombies-2 because exists
zombie-shooter-3d downloaded
city-battle-war downloaded
ignoring assassin-ape-open-world-game because exists
ignoring because size is 80.100000
101-skateboard-racing-3d downl

### EDA on amount of classes per app

In [5]:
import os

total_classes_regular = 0
total_apps_regular = 0
print('Regular Apps classes per app:')
for directory in next(os.walk('data'))[1]:
    total_apps_regular += 1
    
    classes = sum([len(files) for subdir, directory, files in os.walk('data/' + directory)])
    total_classes_regular += classes
    print(directory + ": " + str(classes))
print('Average classes per app: %i' % (total_classes_regular / total_apps_regular))
        
print('\nMalware:')
for directory in next(os.walk('/datasets/dsc180a-wi20-public/Malware/amd_data_smali'))[1]:
    total_classes_variety = 0
    total_apps_variety = 0
    for variety in next(os.walk('/datasets/dsc180a-wi20-public/Malware/amd_data_smali/' + directory))[1]:
        for sample in next(os.walk('/datasets/dsc180a-wi20-public/Malware/amd_data_smali/%s/%s' % (directory, variety)))[1]:
            total_apps_variety += 1

            classes = sum([len(files) for r, d, files in os.walk('/datasets/dsc180a-wi20-public/Malware/amd_data_smali/%s/%s/%s/' % (directory, variety, sample))])
            total_classes_variety += classes
            print('%s/%s/%s: ' % (directory, variety, sample) + str(classes))
    print('Average classes for %s: %i' % (directory, total_classes_variety / total_apps_variety))

Regular Apps classes per app:
flashlight-tool: 3493
stickman-and-gun: 2748
angry-sea-white-shark-revenge: 7367
tv-remote-control-pro: 1837
trigger-down: 2101
amazing-ninja: 2426
batman-the-flash-hero-run: 7326
zombie-objective: 2671
gun-shoot-war: 6839
nob-s-world-jungle-adventure: 9633
goku-saiyan-warrior: 7273
sniper-shoot-war-3d: 12267
bunny-run: 6288
super-oscar: 8147
city-battle-war: 5056
ninja-revenge: 9110
zombie-diary-2-evolution: 8579
amazing-spider-avenger: 2091
shinobi-war: 4303
zombie-frontier: 8146
modern-sniper: 3916
dragon-warrior-z-doomsday: 7380
4636: 9
killer-bean-unleashed: 5654
top-celebrity-3d-fashion-game: 3987
transformers-legends: 4613
guns: 8796
gunship-battle-bullet-train-3d: 7529
angry-lion-attack-3d: 6395
angry-birds-rio: 7272
pokedex-z: 4270
adam-and-eve-2: 2110
jungle-dinosaurs-hunting-game-3d: 5525
don-t-touch-the-spikes: 3801
reckless-moto-rider: 858
crazy-dentist-fun-games: 15772
benji-bananas: 8676
super-tank-battle-mycityarmy-remaster-90-tv-game: 5378

GingerMaster/variety2/222bc1056fbae2a4f1e9355483502edd: 336
GingerMaster/variety2/dd8ea3ae4854f1a878ad89e23b2d3e23: 336
GingerMaster/variety5/9b12b63050eafb36a0fed4f44aa1b748: 169
GingerMaster/variety5/9313fa19ab5d4b19a5c02eee44068375: 143
GingerMaster/variety5/dd06ba9a1dd456bbbad888bca216234e: 140
GingerMaster/variety5/4c1b964b290849a2332550797abda240: 128
GingerMaster/variety5/b81f568fec464f260d988bdbbb2c063d: 628
GingerMaster/variety5/aa44240b167e1111484ba430360595aa: 172
GingerMaster/variety5/30c637fe93a1b9b4aa9047c6ec981cb7: 287
GingerMaster/variety5/4fd35916a26341a1e55bb3ce253ab48a: 195
GingerMaster/variety5/ab3f5567a9f3a310ac20e0a27dc9610d: 66
GingerMaster/variety5/cdd055be9c54df192fdc7d87997a90c1: 726
GingerMaster/variety5/7849301b637e35e6ada8b9a6bc7f6f0c: 1650
GingerMaster/variety5/0e23aa08241ce68912d72e98d4f76ee4: 170
GingerMaster/variety5/b34c39a7e5830a8194311786dd44236a: 237
GingerMaster/variety5/f232db01feca8ea5563a77ac20d33f18: 110
GingerMaster/variety1/779925c209ddba5d0d

Youmi/variety1/7b708c7070569d0d835b8e5a661d7a7b: 463
Youmi/variety1/a7b54e2363d82cd7fdc483469b88be98: 180
Youmi/variety1/f38922d68e7bc14f7c9c4da284833ad8: 1239
Youmi/variety1/6e94d7fe8f06e66a4410900003d9a40d: 632
Youmi/variety1/2ed88701cf10c506f1b7ea7b5db308e6: 2146
Youmi/variety1/e8b75b40e7eec2791f09abb1eb87deb5: 4763
Youmi/variety1/49d43396403c915b65e806fec2fd09fc: 1336
Youmi/variety1/53fdd4717ffa1e755358ee60d9b9d5ea: 239
Youmi/variety1/f62724cddb173f4f57b367c2b79c32ed: 171
Youmi/variety1/3a8e41b57184800c5fc64cd777f61cd6: 1396
Youmi/variety1/322b86018c55e93b8c8e0f23e17f028c: 5248
Youmi/variety1/57baaddb97967496d2752f173fe77310: 1693
Youmi/variety1/ce915c6af4787c9ead701c152518d1b9: 956
Youmi/variety1/bf20da59f5d420cac43a5436e57d639e: 1263
Youmi/variety1/496360a2824f8efb0d2d72b3c30ba718: 495
Youmi/variety1/9028ff6b0d19f29edfa43ebe62ff950b: 975
Youmi/variety1/3cb2dc7ab43fa615a629be244859c311: 1389
Youmi/variety1/784c50f73c38f07b06242354cb1aa524: 296
Youmi/variety1/0bf6bef249ade12b1bb047

Youmi/variety1/75747556f640986981722e3669c4d5ec: 1680
Youmi/variety1/bf2e8ebb752624d0a291f660c8ffc89d: 2047
Youmi/variety1/9fae4b2a00eca5be973283d265b1171c: 1693
Youmi/variety1/3b2047865ce015936eac328f55b3c7d8: 1693
Youmi/variety1/b9916f0c0c731571e3fa4b16d42a338a: 2248
Youmi/variety1/4d45bed4a39d51098a537cc0c334d674: 1863
Youmi/variety1/0746c882cc27e12f032f947306b84a47: 470
Youmi/variety1/e65e31e8d6a619d4473ec2ef8983f72e: 841
Youmi/variety1/317dc6cf9b87168bd0321110ca34274b: 313
Youmi/variety1/0e2c6208278fd19474d0f12a77ce113a: 116
Youmi/variety1/67338720f1bc91be5ceb4c81aebfc0dd: 1162
Youmi/variety1/d92fc2c5866c2059bf34141527a0e876: 1693
Youmi/variety1/e7218b7e64d18c111b032c4247c50357: 628
Youmi/variety1/bd243ba3bf06bf1cb15f604498885851: 513
Youmi/variety1/1ebc7a41c59e56692bf17916662a3c35: 467
Youmi/variety1/33f981315b17ba21e5acdd16bcdb5125: 810
Youmi/variety1/9640c8d9b29b7c74e3cc300912259a17: 197
Youmi/variety1/871765bc88604a468c3d12fe9059a898: 2562
Youmi/variety1/e3c53b406a851a7612a601

Youmi/variety1/e1b9d155f94f23ef903bde2c939a9bbe: 1689
Youmi/variety1/e2455355cdbcd875c2fc16c8db5e1da1: 172
Youmi/variety1/4aca213e939fbd07e041e7801f01def2: 896
Youmi/variety1/253fdb4c13380541d00297cf9b456243: 1561
Youmi/variety1/f78379dc01f13f6220749dd3d2c0dae3: 534
Youmi/variety1/1ebf010c389052cc95f3e4da33f480ea: 1828
Youmi/variety1/e0f612e83e7dd6bb35a5dc3f000228c5: 5208
Youmi/variety1/4b94dafc8ad8e6dcb3e5f06f4ea205c4: 2424
Youmi/variety1/10615b6961762db0ece1ae88573b982d: 1687
Youmi/variety1/0ddc819c4cd623f98a7b6bb477ace624: 1693
Youmi/variety1/5a384df715064e6e44d17d1d5fe6d0e8: 640
Youmi/variety1/fce490a108c46b964d061da545340d78: 183
Youmi/variety1/9ca4a773483fb5b84717ffe21f933994: 517
Youmi/variety1/3f5588b2780b94fb5ce57ce2b3f1c03e: 1641
Youmi/variety1/a5dc3549fa05c5e33947efc9465ae357: 966
Youmi/variety1/4e0bbb9913f3fb396552ffcd994d26cc: 121
Youmi/variety1/aa183df3c9f269dae4fc716558229474: 437
Youmi/variety1/c3d2e10c55971c67405409942787ed39: 1768
Youmi/variety1/10993e2b875078c018d86e

Youmi/variety1/9b22c340156145b481758ef49a7322cd: 1602
Youmi/variety1/185152456f8206156bec6efaa14862ec: 612
Youmi/variety1/ea8e2437c8f8ccc8bff091894bcb43ef: 383
Youmi/variety1/38c2b703cf0e3bf2e4fe4e96f88409c1: 312
Youmi/variety1/4c67c497ed36b51399fcac3b3925a9e1: 1693
Youmi/variety1/f0cb13126d552b37d4e8bfe270ed5b50: 371
Youmi/variety1/e7361da5d60e817a4639aca4f43b7ef8: 260
Youmi/variety1/5280316461aba74cffe8819ab4ec7acc: 1257
Youmi/variety1/66eb9c809fc5e657669b874ee5011ab2: 223
Youmi/variety1/ad47775a4822d3796c697c7e57c40cb8: 777
Youmi/variety1/3cd0e0787ba6fe1bb253eb5c527e5254: 4132
Youmi/variety1/daae5f770f7fc9c9dc49cd7942204a34: 588
Youmi/variety1/fd3f49c38abdf081fe8870e871d937b8: 995
Youmi/variety1/951c85b7a4c46049bf471471b3fc83ae: 181
Youmi/variety1/505b974a04fd3bed420493fb637b1f9d: 1693
Youmi/variety1/17bbb8e2b3d4e97656283f02b84b4c8d: 855
Youmi/variety1/0cb1b2d62fd7d885adf001c45ccbd40a: 1811
Youmi/variety1/3494a20b683b4d49c436720df03931f8: 271
Youmi/variety1/b4fa35556a23a610e88d43fad

SimpleLocker/variety1/334e343769650cba2914b409255ab6e8: 30
SimpleLocker/variety1/734b5f120ad25e3f1c81365f8339d0f8: 16
SimpleLocker/variety1/ee82c6bb3de77a64cf4ec6a8221fc42e: 16
SimpleLocker/variety1/7ed3ad3b72049592dbfda60eca3598e9: 30
SimpleLocker/variety1/63056bff52858f998bca88255065f93d: 30
SimpleLocker/variety1/9a8fc0399e3d2bf87f7a1d6494591f96: 40
SimpleLocker/variety1/93da258efc5d2cde4edf8bb6cda38744: 30
SimpleLocker/variety1/51af8e3aac7a55b2e1f82bd7ecef52c9: 30
SimpleLocker/variety1/94e0d162d0b4d6fdc1b9cb8f8127d136: 30
SimpleLocker/variety1/ecb1b3ce66a394ddbb57c400f5bffefe: 40
SimpleLocker/variety1/a153fc585e4814671599e2533a194a6b: 30
SimpleLocker/variety1/a8200c5ce3a0e1cb3f8fe2427c77a6ae: 30
SimpleLocker/variety1/a995e12ab69d06f1d941030bdc4f4c7a: 357
SimpleLocker/variety1/264ec3c50bfe21fbe425fef21e396346: 30
SimpleLocker/variety1/84f08bad0829c196ae0cd55667b87d5e: 30
SimpleLocker/variety1/9a578b25de159711191e5376510d5a78: 19
SimpleLocker/variety1/139bb7cb84dc74c3e2faef91af024752:

Boxer/variety1/43166d48c5a44132be8e6376e39b3a15: 60
Boxer/variety1/b7823bac1481a66151ec569820904aea: 60
Boxer/variety1/61b4ed5a03eab476d8b47b56eae141e6: 60
Boxer/variety1/fbcc880a07660e9fe4059ff989f9d79d: 60
Boxer/variety1/e4e6911b104f25d846724cbfbf56102e: 60
Boxer/variety1/0d79352b06d4fb0b7cad1e51580cad30: 60
Boxer/variety1/62894cab2c4ac3f60311e4999b9f0877: 60
Boxer/variety1/46399169a6043c34533d0aa765d9cb7d: 60
Boxer/variety1/e24e83c821c49e8d487451240c02231c: 60
Boxer/variety1/ea3781fc396c80f41f529ed4c32b4ec5: 60
Boxer/variety1/27c73e27b951d5af2fff8cbbb7eacf65: 60
Boxer/variety1/0d227b1977799a1e0d41cacf2d3cc2f5: 60
Boxer/variety1/36abc42f5e0abbcd9318bbd60f9f9297: 60
Boxer/variety1/64506af204aee8c27a08f3c323dff3d7: 60
Boxer/variety1/0870bb1526ec47303255d224402abb98: 60
Boxer/variety1/a2dec506ee7f198de3d0a7e5fcb1c5d0: 60
Boxer/variety1/60909989a9c4468dacaabd478a07b1df: 60
Boxer/variety1/855eacc822a78fda1d7eeb7016a30fac: 60
Boxer/variety1/f487a265b966e06b78d13c657953bc7b: 60
Boxer/variet

RuMMS/variety1/7d1c2d11a9b68a107ffb32c86675d8e9: 1157
RuMMS/variety1/859e9dbcbd0db577ff401537ae560e74: 1090
RuMMS/variety1/e945a8a139388abb28b6cecdc8376024: 1090
RuMMS/variety1/e631c06c43899143fec9a5cdc3e04a5e: 1482
RuMMS/variety1/db0df1bddb5485835c76846ea7e6e30d: 1157
RuMMS/variety1/ac6a922fd8c604eb56da5413c2368be7: 1157
RuMMS/variety1/136596d31cc798e9e2a497e955ea7fd9: 1157
RuMMS/variety1/4ed59658844835a222e09c6ca5701bf8: 1090
RuMMS/variety1/b57618a7098fa9fcc14b8779b71ba62a: 1090
RuMMS/variety1/4aa78398d9a927d2c67bf6a5fb0c8db8: 1090
RuMMS/variety1/f20c0284efc5a53b40dea9aabe408933: 1157
RuMMS/variety1/ef61d50f8394b46f85a2f0a750c0b37b: 1090
RuMMS/variety1/4841a521f95ea744243566cc69904bd1: 1090
RuMMS/variety1/33ce815d90ed482a18e162463018f488: 1157
RuMMS/variety1/a3aaff686bf34d60b8319ef2525387d3: 1090
RuMMS/variety1/48378433f79ac304d0bb86ee6f99958e: 1157
RuMMS/variety1/d1c823af96c52b38cb4330c328a6727f: 1090
RuMMS/variety1/778cc7e83ad27c92f30cea519989f47b: 1090
RuMMS/variety1/851953bee7687

RuMMS/variety2/5144790d272daacc7210fc9e2ae41f12: 1459
RuMMS/variety2/18d1b845b2ee1960b304ab2fd3bfe11b: 1459
RuMMS/variety2/0c3dbcffb91d154b2b320b2fce972f39: 1459
RuMMS/variety2/e5e8169fa5aa58c3e4ca1b93a807cd1b: 1459
RuMMS/variety2/6148b71d713c80af2acfd3506d72a7a4: 1459
RuMMS/variety2/93323852f58c4e1b436a671651cc4998: 1459
RuMMS/variety2/c939bae3b894ae42ecc249eca8ce7a74: 1157
RuMMS/variety2/67fe7190cefc9dad506ed3c1734ff708: 1459
RuMMS/variety2/f9a5e29ebef80138f9b59de9918a3965: 1157
RuMMS/variety2/640e5a25191af807e3103edb26f1dd1f: 1459
RuMMS/variety2/e5582e727ee6f65e5c154cbfc0ee3158: 1157
RuMMS/variety2/66e9dca8bb42dd41684c961951557109: 1459
RuMMS/variety2/8b56c493375d3b65d509793751509ba5: 1157
RuMMS/variety2/a4f19520957bee3d68755a3978fb16be: 1157
RuMMS/variety2/c5dd6c26c4c1e03fd1ec51cb1dec91ca: 1157
RuMMS/variety2/65eab2ed600f5ae45fe916a573ce72b0: 1157
RuMMS/variety2/ba1c5315933c1a4d446bf90eb9d7c8c6: 1459
RuMMS/variety2/00c7f8706c7a2313e910d686e7f83ed6: 1157
RuMMS/variety2/a893d3cfce6e8

DroidKungFu/variety2/36f3c3c0b76b4549fc7336ac112189a8: 353
DroidKungFu/variety2/2dae4633600d6fe41f5c3e739cce16ca: 69
DroidKungFu/variety2/ccefa1eb4c3ec91e5becbd3076288d48: 80
DroidKungFu/variety2/d7e92a44e606d6da95879612badf9178: 69
DroidKungFu/variety2/2cd362f7f4a5041af5642c6af6cfa59f: 337
DroidKungFu/variety2/68cc8cb6b42ac3c121d6bb6d59ffd0e3: 86
DroidKungFu/variety2/00268453be254779f0c7590de47db944: 1072
DroidKungFu/variety2/9fb6aecbafa856ce7381e842285af6a0: 353
DroidKungFu/variety2/e20c743fe8582cf6065b47bcf780a1de: 203
DroidKungFu/variety2/54d62e3205e7077d7992d412d81bf767: 243
DroidKungFu/variety2/7abc319ed7c8e663edab7a5760e77f6c: 1072
DroidKungFu/variety2/e2db6cd8b7552c231df69f53a4a77f03: 163
DroidKungFu/variety2/c69e5663e37740b279e44f122968ae4b: 69
DroidKungFu/variety2/ee704545a7b132dcc08e0e4e52ce8152: 255
DroidKungFu/variety2/85a8e874d4bfc5f7d18fb4c07b9fdc4f: 330
DroidKungFu/variety2/c4e0546177dfc88603394a8e2b3f2f04: 176
DroidKungFu/variety2/cf5f9633de328e10cda8f12d2e8c4161: 1072

Roop/variety1/fdb946b5fa77cab9e98bb992080f7600: 125
Roop/variety1/09f72ad8935003bd842c118ae5c22501: 125
Roop/variety1/2b76b9694b7aecf49bc7486d6994f409: 125
Roop/variety1/921d97897142334c78d4171ac4d16cd4: 125
Roop/variety1/5d92b64832e2ec0f98049fd595f7b804: 125
Roop/variety1/7f96188d1bd2905f8629365171cfba10: 125
Roop/variety1/4f6fc31adc277e813fe103097422cf7d: 125
Roop/variety1/788b0f9b1cac0cc253cfe0774c660dae: 97
Roop/variety1/85ca9119a8ab9ce99cb8002328d738d8: 112
Roop/variety1/d49d170de090fbacb558d4444cd54aa7: 125
Roop/variety1/73d54e708a321924c4b6bc8309aa3851: 125
Roop/variety1/101bda93b50f08fe1bc1518f9f2a3aa2: 125
Roop/variety1/bdb8524b454ab505cda96a656f6201b0: 125
Roop/variety1/f96403cfbfeaf444e8e7107add7b0031: 125
Roop/variety1/432c66f870467998e7bf9c29d42ec7c4: 97
Roop/variety1/91530cf3db534f5b42ba26682788cb5c: 112
Average classes for Roop: 120
Vidro/variety1/b582078efb4c5f60a0d037602b9ffc09: 100
Vidro/variety1/f1ee442eb94340ff47fa0269e393d0e0: 238
Vidro/variety1/f8031a28cbee3b7a086

Koler/variety2/790b65d3fba5e125068602f48f5791b3: 1413
Koler/variety2/d380b83516975ffe4c140d0abaa77cb8: 1413
Average classes for Koler: 133
Lotoor/variety15/9dc1a6770bf57142bbfb9428b3e6213e: 326
Lotoor/variety15/1edecf2b3dc5a702cd5052065a352d38: 17
Lotoor/variety12/fad81626c3dc0d7af60d58b434c11d3d: 1643
Lotoor/variety12/f37c79437efeb2a8cc46c382c63c3943: 262
Lotoor/variety12/f852e9de2069611f5b047867e49059fb: 119
Lotoor/variety12/fcdfcea712c8c3e8173220ee0088f5c7: 1848
Lotoor/variety12/fe14819f6770bb249d87a5e2efe442c0: 122
Lotoor/variety12/fc9713f464449159adcd6a515fd02b07: 119
Lotoor/variety7/5343038848a94fe880926c1e2b4f46dc: 128
Lotoor/variety7/18ee75d03d2d67e869059c87499fd6c5: 19
Lotoor/variety7/12714aac557bea7c8a31cd4dcc0941c8: 115
Lotoor/variety7/bf51b46bbdd4d9565beadfd868314002: 25
Lotoor/variety7/baa8ba9a36a11c9daf52e7e6e88636db: 25
Lotoor/variety7/4634f88b8dfdb2332b447c0d85ff4bb6: 24
Lotoor/variety7/3c59630afc362dfad79292a9fb66274d: 162
Lotoor/variety7/ccfd2ba0fce7a600853afbb89d2965

Lotoor/variety6/86e0ea6f8bf076bf37258c2028fa1980: 478
Lotoor/variety6/980097d21a345f10a27260b09ffe462c: 8
Lotoor/variety6/a03b3d7daba51161610a7566732defed: 8
Lotoor/variety6/69fa7e3e8542ee83dfb1de6a4a2dbcc4: 8
Lotoor/variety6/d4cb282116c10df5c0d57fadb9ead50d: 477
Lotoor/variety6/1b9d1628268a9e5ab9c1aba2e4800c2c: 3
Lotoor/variety6/bc588ebfc9dde76ca0933f7ed6b36ed6: 0
Lotoor/variety6/4121446a412a622dfa6a751a3278fd43: 0
Lotoor/variety6/9a818b3ad43a40a13cb3d6a101f6acca: 477
Lotoor/variety6/683518c953ad9236ca95e48510da0182: 493
Lotoor/variety6/6a6f03a5001ca3b64f83472f87885c2c: 16
Lotoor/variety6/673eaf06dd1236ef13afded303cd1ba7: 16
Lotoor/variety6/58406edc9756685310b9683ef0b88cea: 8
Lotoor/variety6/cf1eca3fdf2dcdcddeae809fecae2a32: 16
Lotoor/variety6/0dfe260bdb43b47dfabb85ab677ba58f: 16
Lotoor/variety6/2683fe3cd724ae0b3b105e4b9ed6dc91: 496
Lotoor/variety6/da335ba83c1ce79971dffddb5a239c6c: 0
Lotoor/variety6/1c5a080f8069c02936ed9643e5c14b23: 8
Lotoor/variety6/cb14aeb47733c17458fe707cf45d220b: 

Steek/variety1/ff28b758f18030c14402e100dbb6987e: 1484
Steek/variety1/0de5c01c9e66be313970cc3af017f188: 1484
Steek/variety1/1407cd7c568576115204697fdbbdfa43: 1484
Steek/variety1/1a4ed1ca65321659b139f9cba9c9cab4: 1484
Steek/variety1/6606e8adad40e3c5b0b8c347a38eb86b: 1484
Steek/variety1/bcb3026536783bc774a05d93bc2f6039: 1484
Steek/variety1/a879ef0f3daa3b66eaf9a713559170ba: 1484
Steek/variety1/98eb1f31945f4cd97088cf9fbc49d03b: 1484
Steek/variety1/a14790b98c0352d81e1b70db8a046aec: 1484
Steek/variety1/d62b2137083cf1d626c096a3a51815fd: 1484
Steek/variety1/c69d0d8b86bf3946ccbc011767b06919: 1484
Average classes for Steek: 1484
Fjcon/variety1/92b123ca701ff2254fdaef70689d4cbc: 357
Fjcon/variety1/b3b31aeeea483cce8b38fdb73a9a5403: 503
Fjcon/variety1/a004bc17a3aae69b7734ce80662e6c9c: 357
Fjcon/variety1/73f75636d00c766c6096e6051e474a76: 357
Fjcon/variety1/1569236acc16fa45e90f5f7ad9221bcc: 358
Fjcon/variety1/2ef46c5a75c2124e0c4fe04884980486: 357
Fjcon/variety1/cfa4f9a03dc57fd0338bf524e0aa474a: 357
Fjc

Minimob/variety1/080ab3c5b9e6305a2a9e996a23f47047: 2172
Minimob/variety1/ddb77869b98f0dd17c806f724f6b1f71: 4011
Minimob/variety1/4902e933dc6aed2958ac112d2064d11f: 707
Minimob/variety1/e9b570133d8084f62b26aa02d22b0df7: 906
Minimob/variety1/c1bbb026b0511ee9788574890f8af39a: 2337
Minimob/variety1/20575816ec50744371234e2065fc68c2: 4497
Minimob/variety1/68cee363bc51b4d5655e0f3c37998e5c: 421
Minimob/variety1/9d0bc383ad25eca5b792e33c6fe2df34: 2011
Minimob/variety1/2e276bad880dd67482fb0edda67dbd4b: 1542
Minimob/variety1/f6e3ecb453d07bbc8764d3d617b08f4b: 238
Minimob/variety1/99cb2b258722cff534b5ad99e26191cc: 873
Minimob/variety1/5923994c02860269a2927a79b990d6c3: 2011
Minimob/variety1/3f0aeb450fe7a88ac239a25415d92807: 586
Minimob/variety1/6ba593146023f1dbc92b6cb291ae0fbe: 684
Minimob/variety1/7bfda83a347d8b492685410d5468f412: 176
Minimob/variety1/65581a10085e5c3d357aa03a8d5c550e: 2040
Minimob/variety1/9da9e2e75b39021f471bba1f9086afdc: 216
Minimob/variety1/430be7f36df70e01ad82908415d6d2c5: 793
Mi

Nandrobox/variety2/b1e077bc19891b6c45eee4df6c741473: 224
Nandrobox/variety2/fefb8ef9e313cd4f7e11fec6fa7bf6c6: 626
Nandrobox/variety2/f0c792e54d6399979d1e49743cd73bda: 223
Nandrobox/variety2/4fe8e3849f1e3ef70e5f0ab74c344aac: 388
Nandrobox/variety2/fab6cccce5f255af0d50b3638bf1de3d: 467
Nandrobox/variety2/1f8106bd3924fd8ddca84b5b646b4cbe: 381
Nandrobox/variety2/4d871daec6a4270a70f42800f9a932f2: 636
Average classes for Nandrobox: 590
FakeDoc/variety1/063846036d6b5d45998879ae2cc142ab: 272
FakeDoc/variety1/e11b2efaf05034a2b3f3e62e5bbb630c: 272
FakeDoc/variety1/41b93df6d41ea33abf9cbb7de4652f3e: 272
FakeDoc/variety1/c1154bc5c619df2609f1f785c8fdbf2b: 272
FakeDoc/variety1/3ab3a1d708811f0e8b36e1713ff9dbfe: 339
FakeDoc/variety1/dc810b5de281d84a654784d32bddddc7: 272
FakeDoc/variety1/8d7a9c7ba8a39f641613ba3e9835d2a3: 272
FakeDoc/variety1/2538cf4626bde75e6bd0e120389bbfee: 272
FakeDoc/variety1/07ee6cdfd8e5c111a33dc6e40aebb3c5: 142
FakeDoc/variety1/af5afc5e135e9a8dbab77ba0b935cf3c: 165
FakeDoc/variety1

FakeInst/variety2/6d60e6fddd3fbdfd4d0b2d6bce25170e: 39
FakeInst/variety2/6bc6e90c77a919fe0568c6e5c4eef562: 38
FakeInst/variety2/b24922f7f9308298b2ac3c5fa9b9e4e9: 39
FakeInst/variety2/20fde0941db01f4707b45e96dcf6f640: 38
FakeInst/variety2/46c902ed57beea3a00a6c2275a220129: 38
FakeInst/variety2/1b4c6b2e39336a1aa3939ee3a9c5865e: 25
FakeInst/variety2/afe0718f1547518ac443b2240438675d: 38
FakeInst/variety2/a2af76eb4215a2e41c60accbb39470ee: 38
FakeInst/variety2/13591a56f65c688de0726f84156f7ff0: 25
FakeInst/variety2/8a00e182e9517a13b4915960dd8cd1d7: 38
FakeInst/variety2/9cc99642dd7712d22e2d2f31e33af90d: 38
FakeInst/variety2/51daa9651d546de844280eaf722c3e67: 25
FakeInst/variety2/0b8275cef802c5356b3cf523668114e3: 25
FakeInst/variety2/0a7a631b5ad0c7c8013adba356597264: 25
FakeInst/variety2/600bcf4dacce1a23617ff7a8809dce81: 25
FakeInst/variety2/c95da3f6cd952583b89e77c446f67b81: 39
FakeInst/variety2/438caec9a44e0f74a54090739861ec7a: 38
FakeInst/variety2/af8eb7dbb0532342e35c3e90693b4a4c: 38
FakeInst/v

FakeInst/variety2/61624fde08ac3c9beb0995840979896f: 38
FakeInst/variety2/142594cca94f3ad406825e35a46e254f: 38
FakeInst/variety2/bf3063ff38de2acb5bf36b8559f61036: 38
FakeInst/variety2/16acdc03e3ab6c0e04a5bb946815938e: 38
FakeInst/variety2/5a3638e7107a82b63625b66f7075d15a: 38
FakeInst/variety2/9dec8b2d2885df9ef4fc5dfb59ec95c6: 25
FakeInst/variety2/382d20db97965b133673f3408a2eb289: 25
FakeInst/variety2/c00f5217e12e037226998a26e0ee32a2: 38
FakeInst/variety2/6cb45cfa7916bd92abbb05aae5bd9e01: 38
FakeInst/variety2/7ea60695b26c51034a7f20fe9f2fd2f6: 38
FakeInst/variety2/0b661aa2c924a3d4965e60a163ac4f20: 38
FakeInst/variety2/818991cf6a830472a051be233fc3d891: 38
FakeInst/variety2/cc1bf5d397f8479618c33aac82cb42f7: 38
FakeInst/variety2/2b449dd48f75a05d84eba9eb4b20528d: 25
FakeInst/variety2/0c87efa6c0b520445f3f498ff56b1e3f: 25
FakeInst/variety2/736b96e7f9857a7fe68a1f21c5034258: 38
FakeInst/variety2/19626b75bed728bf4793930df0297da1: 38
FakeInst/variety2/4d952322a6ef2a153b8ea88e2af143d5: 38
FakeInst/v

FakeInst/variety2/c1ff1ed151f87169b49dc7ad770a54bb: 26
FakeInst/variety2/d83ab6df05930bccf2bf730e3fead9d3: 38
FakeInst/variety2/192f06efc3555715cac6912a120bf83d: 25
FakeInst/variety2/960a09bec3c35c05cb0427041416280c: 38
FakeInst/variety2/6733034d7024308395f0776f9f6e591b: 38
FakeInst/variety2/8add83f650fbcabb421929cb70deca2a: 38
FakeInst/variety2/f20365eb5273ad45fb86394376991206: 38
FakeInst/variety2/61454e8bb6a377f487268591159efe18: 38
FakeInst/variety2/0da2d4c2d2107216f842046b3d2c4dd3: 25
FakeInst/variety2/3722060d01e4dae85d851db708428677: 38
FakeInst/variety2/89caa10dbc1b0b61829b91780a753d0d: 25
FakeInst/variety2/08b8734605ac676af79d5230196635e7: 25
FakeInst/variety2/abe1150c89290199a96d3acf2b164e36: 38
FakeInst/variety2/ceeca24713e1cda6e88094bcae2fe251: 38
FakeInst/variety2/192f0f463b369f3f2a8c5fa7028840a0: 25
FakeInst/variety2/dae5940d932e049007539e96ff8d4ca9: 25
FakeInst/variety2/d3b0da30c386fde4c6c42278dabbc25e: 25
FakeInst/variety2/a7fba3000f1595e0d54ef67e3951b18b: 25
FakeInst/v

FakeInst/variety1/13ce82ff74bde89475175642c1395570: 19
FakeInst/variety1/bac5c0643f2e3f770819ef6fa53a0107: 13
FakeInst/variety1/394d6801e35bdb0d32a35e95c3442def: 19
FakeInst/variety1/264186801487bd509dc979f159ce8bcc: 19
FakeInst/variety1/25a84bb13b8456063c075397ef718d1a: 19
FakeInst/variety1/a2eb44b95f0c3051690ff2b3f685d7fe: 19
FakeInst/variety1/1f2b195619744a277610dde6b268aa83: 19
FakeInst/variety1/8fc2eee87448df0400f5b15ffdda7c55: 19
FakeInst/variety1/2a96efaab45addbeab73abb96ee91034: 13
FakeInst/variety1/184005812b425d24d45e1692cfd66c2b: 19
FakeInst/variety1/018d500977b396d09b66bf9f9a2da85f: 19
FakeInst/variety1/0916282ef1e06ef8fd9d13c4ccac3bac: 19
FakeInst/variety1/6d13378ca4cbf7e5cbb657595f2fa02b: 19
FakeInst/variety1/4c6927fe979216acd05990de5a3233c7: 19
FakeInst/variety1/5d8c23340b8f7e4c71d4bb0d5917b8a4: 16
FakeInst/variety1/26be0fca7122df266c285b725d563f60: 19
FakeInst/variety1/2dda15aecadc70474bb5a9ae7eb8301a: 13
FakeInst/variety1/7fc9d84930deb20e1c9e675a6edbbf22: 19
FakeInst/v

FakeInst/variety3/a15b45541d9840f679bb22d2cca201ca: 48
FakeInst/variety3/03d0947f2a685bea1d1c6fbe7c3ec6db: 37
FakeInst/variety3/54d897f56d46ca8c85604de5ec082e48: 39
FakeInst/variety3/f7bcec3df011a9bb9544159007703d41: 48
FakeInst/variety3/93253d8cacfe25a3913703b92e6a640e: 48
FakeInst/variety3/aabe2950cbc535770d913925e635a45a: 36
FakeInst/variety3/b36362d57016ffa4993a41dd3fe56299: 59
FakeInst/variety3/b9928a64750515349182ab31c2cbebce: 40
FakeInst/variety3/bfd4ea829c9b69d1dfd81d52967065bf: 46
FakeInst/variety3/6b7e36f2590df6bf2ea18514facf6a10: 48
FakeInst/variety3/d50fdc2426d4e057654406e8b12b839c: 39
FakeInst/variety3/b8a485b6869683297a0758d540dc12dd: 311
FakeInst/variety3/f23beb43b12fbd404a9587ae1090467d: 40
FakeInst/variety3/d4116bfeb0207c518adff5359df77689: 46
FakeInst/variety3/a68beb9687768751e9b9d9c7964a25fa: 38
FakeInst/variety3/3d09a2fd1ed6372574114727ddac34db: 62
FakeInst/variety3/ebb497608b9fc1fae158c1928bf7b19e: 37
FakeInst/variety3/989c72d144a2747c87507a97e704fdab: 60
FakeInst/

FakeInst/variety3/7fcb8384a3ddb6be85e2cab75ddde917: 60
FakeInst/variety3/76510fd7b9d517fd2d6649ac30b71904: 40
FakeInst/variety3/8e56980a9f8722d63b41268fb35de9b9: 40
FakeInst/variety3/b8aee3d5ca7694005d38d6bc6ebd7a33: 60
FakeInst/variety3/5221a70314be3ee4a5e80453a1cdefc1: 40
FakeInst/variety3/d51b3e3ec4d0246fe2487e556c4fe542: 39
FakeInst/variety3/bb98afdda813c94309f7955bb858790b: 36
FakeInst/variety3/31671d62efbb8c2b46d486b92b74ab82: 61
FakeInst/variety3/9d95e0d32aa886774e80f220c3113f5f: 39
FakeInst/variety3/eae3a04bcfb09f7a30376452d1bcafce: 40
FakeInst/variety3/fc9d722fb71c17458f1aec3cce3eaa1f: 59
FakeInst/variety3/126217b2b04f4c161735ea775b5e8517: 60
FakeInst/variety3/514e1c6450546db967191ced6610fa0d: 60
FakeInst/variety3/7f25fc418fca7baa1466a066b0cf7243: 41
FakeInst/variety3/c49ceffba09cb7c4d0d5ecafb7a2c330: 41
FakeInst/variety3/5ec5e1626313c79ebe1d574b570a6859: 39
FakeInst/variety3/6671a31862a3af20b08feac6960f14dc: 59
FakeInst/variety3/291e0b2048ee8efbd4b182eab8414530: 38
FakeInst/v

FakeInst/variety3/8822bee8c7d5c7e913d2b8f5ccc944fb: 40
FakeInst/variety3/bf343daeffabd4577f0848ebee94985f: 45
FakeInst/variety3/df8d18a4f391517b4495826cd0472156: 37
FakeInst/variety3/a19957af778d77a04320410cd720346f: 39
FakeInst/variety3/b7c1d3afe06a703d53fedddd688b7eea: 38
FakeInst/variety3/2b2fd4635271672e739062dff2e5dcc0: 59
FakeInst/variety3/9cceb80a7b3e872e029273ba47941fa3: 37
FakeInst/variety3/f98fc5b9298de26855d697c402f85c2b: 40
FakeInst/variety3/7744a6f0df0c3dd8de12efbaa24a6c5c: 59
FakeInst/variety3/5a6cd0ff0d12b423e0efe0c6fc83bc02: 39
FakeInst/variety3/3d17fb4b310941dbb7ba38af8babbafe: 60
FakeInst/variety3/c450d0c63d5ea756c1aba3f9e6e866e3: 41
FakeInst/variety3/11e9fbf897a1327b621fa103ebead210: 38
FakeInst/variety3/823393380bd3d51d9943da52c29b061c: 40
FakeInst/variety3/0875a513771bfd36668010b7d4e1f788: 36
FakeInst/variety3/4f3e07fec0442af1fd1d121fa7a46c13: 37
FakeInst/variety3/38db67531efb97a9883a865f39fa990b: 59
FakeInst/variety3/67ef4fb4a307fa77f34edb9f150af2d9: 39
FakeInst/v

FakeInst/variety3/997b1a9e40a80913eb5869a755a440ea: 41
FakeInst/variety3/8516e498066e58dfc138be05595fcceb: 40
FakeInst/variety3/26cfdcffddef3a653ec7a3775809a3df: 48
FakeInst/variety3/d8f4ef7d96be55158f568d5705d8b1fa: 59
FakeInst/variety3/a59d1ca6678b42945e37504b4380c3c2: 59
FakeInst/variety3/2ffa94ebb411731375cd810fc52ea7d4: 39
FakeInst/variety3/e36e6c0cff24e41f82faa869e5b56d01: 40
FakeInst/variety3/8436f77350ce02ec8f76fffdfb08a4a8: 40
FakeInst/variety3/b83df8e33875745aaf2459dcc9f302b2: 36
FakeInst/variety3/d2b2e1c42839525395d0e41b37f68779: 39
FakeInst/variety3/48abb9da11ee38ad748d2b1f518ade55: 41
FakeInst/variety3/42bc9e3d01f8157011af66729c0db0af: 59
FakeInst/variety3/5d22bb974bede87b5c3cccbfc446f598: 47
FakeInst/variety3/bad167c3b453604fb703ff41799220e7: 48
FakeInst/variety3/8d394e309f7bbc3ea208c80197c69e8f: 39
FakeInst/variety3/71b9b970b7242a4573bfa8eddf71c9e8: 41
FakeInst/variety3/9ca9b22352c80d191ff4ea0685e18cb4: 59
FakeInst/variety3/beb7e2b47002f120b2167580ae9609a5: 39
FakeInst/v

FakeInst/variety3/0714787defa3bf535351209e3abde7eb: 48
FakeInst/variety3/c877ce1d49cb72beb2a9ebe224b288a0: 311
FakeInst/variety3/adc31607c8edbbf5a53b5d8797b5bdcf: 36
FakeInst/variety3/8955fd0172604b2fb4a4cad24bc10fb0: 311
FakeInst/variety3/c313a126e12481948cca76d5fbac1ea4: 39
FakeInst/variety3/627603dd166e05c2cd500f71ab49609e: 82
FakeInst/variety3/4d582975ec979c49c309312c3f8b224e: 40
FakeInst/variety3/ff51eef2087c33e02adee425e8e7c625: 47
FakeInst/variety3/7db99825f4e479ee47641ca38033e459: 59
FakeInst/variety3/c114dfd7f364496fe64a4c0d210393fd: 40
FakeInst/variety3/ca3b435de7c6d0d7d58aaa0f939551b2: 37
FakeInst/variety3/608893b135026f1e081f5d5da189c102: 48
FakeInst/variety3/a342551b4ce2980f2df57a8c528fb5fd: 40
FakeInst/variety3/162c1add80ed939e789075aa583bb3b6: 40
FakeInst/variety3/7e59b58723e05bf8e759137aae7ba18b: 37
FakeInst/variety3/dd6dc297a7373259a494e8f4f3df98fb: 39
FakeInst/variety3/7bc81aec31a16e706d3e7c33956090c7: 46
FakeInst/variety3/5283253f7907706d1e5089a7d251c4f6: 40
FakeInst

FakeInst/variety3/92aff6b99d6edf8741780c1f344484a7: 59
FakeInst/variety3/9b87b2315d737eb39c50bb544820891c: 39
FakeInst/variety3/eaaae799644a38b63d4e697951a55f30: 41
FakeInst/variety3/2fb6b2306306d54bb8e44e514fa380ef: 60
FakeInst/variety3/3b5ab33f68fbbf9465c8f74767743ffd: 39
FakeInst/variety3/d93ca860d83136a8469c961d7211b56b: 41
FakeInst/variety3/56369def7c111ff7256dba26bea9ab78: 39
FakeInst/variety3/9b0d9ea4bb2a5549f5300595b1eb643e: 39
FakeInst/variety3/fd6782b94386c19da9ff487d7c89254f: 60
FakeInst/variety3/a971143f9c23db05b3d7602efb9e2a98: 40
FakeInst/variety3/e984a5ab9cd235bcd704a7f6f7eed775: 312
FakeInst/variety3/8afb7a5199b46db5ddd5345be0c6cefb: 40
FakeInst/variety3/dd80e8836b864c60a20e0650aeec9904: 59
FakeInst/variety3/00a4d3a5760a79356a545e68def552dd: 60
FakeInst/variety3/ab2814b13b5e1e78655780dd0a1e9682: 41
FakeInst/variety3/5d2a50ac62e969f023f998cc1d66bb19: 311
FakeInst/variety3/fa2a79e9cf4696b01e27b116878003d2: 60
FakeInst/variety3/9929c87677f48401dfbeac29978b0cb6: 58
FakeInst

FakeInst/variety4/caa9a7f958e6b184fd292b55a2774b07: 34
FakeInst/variety4/4a8c3560f3393128b77aa4906df0e9fb: 34
FakeInst/variety4/697eb006194d73e9ca0db6e3c1db8a1e: 33
FakeInst/variety4/dbf68991202b695852cf074b066a8148: 33
FakeInst/variety4/181a8dd8369a03a197b6bf6e0a443ff2: 33
FakeInst/variety4/6e158ba396f3be4f6a2d2ef482000f44: 34
FakeInst/variety4/e9cfbff401f36c20fd4b302ecfb0de00: 34
FakeInst/variety4/9df830a601235391c6b0be7927cf737b: 34
FakeInst/variety4/3f8746cfdc6ae1fa5276d0d77b3222b7: 33
FakeInst/variety4/2155a9a983293dfcd26146d08dfc372a: 31
FakeInst/variety4/cf1e9ddbfd2a75915741dcb296b0da60: 34
FakeInst/variety4/4057acb51dac56efdd2376bfa29f62c7: 31
FakeInst/variety4/de681d96acf5cb5b5c4fe74e894b2a12: 33
FakeInst/variety4/052da738844d37e22b76685f4a4933e4: 33
FakeInst/variety4/98f0eaf861b2a3c32945c0a37f66661f: 33
FakeInst/variety4/35036b1b46a0708878e4669e8e69c8c4: 33
FakeInst/variety4/052d742c7ccf3c6470776187d2d7dad2: 34
FakeInst/variety4/da28ff01aa33b9f05db45551a14f70ce: 33
FakeInst/v

FakeInst/variety4/622902045bb6fe08dcb33166db79ca5f: 34
FakeInst/variety4/6412b42eee7cc3959e894d6b478b3195: 31
FakeInst/variety4/ece2f86b3b565686551511e52b3d6384: 33
FakeInst/variety4/51c9d5bf36d70fb4200f859e913f847f: 34
FakeInst/variety4/e29be48dddeaae9663b995b3d39540fd: 33
FakeInst/variety4/ba0f623067529c7d3b307a71a2d180fd: 34
FakeInst/variety4/0144d5d82a40ac04fc9b1b665948197f: 34
FakeInst/variety4/28dfbfa3cc26b488d413e48e34d226e2: 34
FakeInst/variety4/7c09537fd3a638a6011caf9a7b193fed: 34
FakeInst/variety4/47626e1be74e277511c8247c1f295209: 33
FakeInst/variety4/141034dfcbbb87aa2914fd2481d28326: 34
FakeInst/variety4/ed30776ed1cbcc62cb9e19860b31d068: 33
FakeInst/variety4/2da470f7e57f9e13bfb2d3f08bca8aa1: 34
FakeInst/variety4/d793c4a4a6c91a51f48a10ba0ce7be12: 34
FakeInst/variety4/a2c4188b7d01f382eee3243e87b7bb10: 33
FakeInst/variety4/773f777eda63d9e5213cb2f691174abe: 34
FakeInst/variety4/a104d231fea4793c0a2330d4ec7ccb4a: 33
FakeInst/variety4/87f67162d45b6f814952ab26ed70e839: 33
FakeInst/v

FakeInst/variety4/14f8763a572d65c968721c78da76e5c6: 33
FakeInst/variety4/b19a706845bd1f17447397ff4917a4e7: 33
FakeInst/variety4/7c7a9cd2a531d49998f6ae31995e037a: 34
FakeInst/variety4/1e5cf854c2d37086466bf20b3957c710: 34
FakeInst/variety4/779c73b9955342456f465002618a8be2: 34
FakeInst/variety4/4c92a343c19e68424e7e0d2a90c3a967: 33
FakeInst/variety4/389f0331407054adb62a0d292a51b6bc: 33
FakeInst/variety4/1f01e1cbff7f5793aa56a1bf64a89590: 34
FakeInst/variety4/f9906adcf50a91d5c14b83b7ad8983f1: 34
FakeInst/variety4/e5d5760073f29c5281939b2204f5f8df: 34
FakeInst/variety4/7cb9dac9518f27b6239320d5b9d1912b: 34
FakeInst/variety4/bf6f2fa6792661c5da91ceb66c5996fa: 33
FakeInst/variety4/40285e00b0e9204bd25cb7dd34d1c2c9: 33
FakeInst/variety4/e811c29a0bf53ecf3af2b4254a73384a: 33
FakeInst/variety4/4fdcd751289c004196dea4eb1d17de3b: 34
FakeInst/variety4/8acbbee055cd026acfb66ce1c67e6a83: 33
FakeInst/variety4/622c1e4587b13d21af549c23276cde86: 34
FakeInst/variety4/a148ed61d84a2d00c0a903e10d84d12a: 34
FakeInst/v

Kyview/variety1/dd7e61b816842c8f28a9c4888a19ad10: 1751
Kyview/variety1/eba89becb7ade6ce1c8a75c83eea5b7a: 1000
Kyview/variety1/a21e99be4ee8c87da8432ecf3c54298d: 811
Kyview/variety1/9aa0e049a73532ea3882291c190bd5da: 1271
Kyview/variety1/b2d4a8d51947cf3f1b2d3f05ea936777: 1065
Kyview/variety1/4aa6ced61da15c18186b17e9af729a46: 811
Kyview/variety1/fb30a38d057ca4fc2dd92277522d84eb: 1539
Kyview/variety1/2643602b4d491d5e841142cd103a87f1: 888
Kyview/variety1/da240bc9b854909e9a1bc047c603496c: 1643
Kyview/variety1/c9bc815dc615d7771418a1f651a8a902: 1138
Kyview/variety1/5ae44ab11eb3691b642f60477686e4e5: 3835
Kyview/variety1/e7c81a0f3289896a99cda9c5f3720f4d: 811
Kyview/variety1/ca56f5048df775b15724f6d56ddbd9ea: 1019
Kyview/variety1/075299a43b49e1fd63a2093a7ec8959d: 453
Kyview/variety1/c715f5ff7e6633b3fb11e20e01b964e5: 1689
Kyview/variety1/4e8ef6621b08090527fa6864421927c9: 1020
Kyview/variety1/dcb1fdba68fed475081c2e34579794bd: 1604
Kyview/variety1/f136efc2985a5deb477de0193de6a9e4: 1701
Kyview/variety1

BankBot/variety7/084ba13ddcf17b675ee86b6e85f7cabf: 609
BankBot/variety7/8bc6b49609b0c595aa97b8f31f86c645: 1415
BankBot/variety7/319fc149dd63e02ffc8021afd34bd410: 605
BankBot/variety7/2030a28c3a874010bd73bb92cb95b4a1: 609
BankBot/variety7/33265aa59b03965a16352b229170eedc: 609
BankBot/variety7/2b08fdd40a9d99aeea8af7cd92ac6752: 1415
BankBot/variety7/2d8991403092408e5b7a03e89ed7317d: 603
BankBot/variety7/9add025feea977cf18f9d0236e53f108: 605
BankBot/variety7/bc286727ff5a38e2c9d4c6979e2cb296: 1415
BankBot/variety7/13b23f093555add17df70ba7df560c46: 607
BankBot/variety7/371e1dadaf4eea5fdaece63a080b6b9b: 1415
BankBot/variety7/c2f64d7f971ff479eb7cae06c8357ee8: 1415
BankBot/variety7/5bf8dbd812d248b8d510e35d27617458: 1415
BankBot/variety7/58229dcf4519f930f44c908d1c683071: 609
BankBot/variety7/12e4fd07dd3b0889ffa57e81262d515c: 605
BankBot/variety7/3317331e75aed00e9a26a7bd5cf83236: 613
BankBot/variety7/f120d71ebbfe390ecb84015248961069: 610
BankBot/variety7/ee8a99edea4053543dcb07172851cae1: 609
Bank

BankBot/variety7/ecf37d543490bf2875b808d6d97f904f: 603
BankBot/variety7/3c1e5c7916714ed442ffd7a27165cd40: 609
BankBot/variety7/d0525716bab0223807529a440ec04e23: 609
BankBot/variety7/7fb313a0f2ec97466d9ef493d4af41f1: 603
BankBot/variety7/1701c97b0ebf3cb86500ed8ca2decf43: 609
BankBot/variety7/9bb0c1e675e372a76061f958ada44c77: 1415
BankBot/variety7/98eab6743d32f8fa2992f05744a372ac: 609
BankBot/variety7/b91c57c8fb93bf1d9f54591287b07239: 1415
BankBot/variety7/43f5c943ec2ae45da0c43a85c387b892: 1415
BankBot/variety7/214b4adcfca01a264670d1a0387b01b3: 609
BankBot/variety7/731b6c3b289aa257341c9258c7994071: 605
BankBot/variety7/02cf263dd0079ead46a7dadb8835d3ef: 607
BankBot/variety7/2c7f1df2587bed4b82e2216591fcdb66: 603
BankBot/variety7/d0c79041d545347b29e899d54d3d6106: 603
BankBot/variety7/ef88f08d4296b3660e7b7b0e63dca626: 609
BankBot/variety7/f2762eb506c9cb7a2755f76ed3cb1901: 605
BankBot/variety7/6cf089fd36cd060dc59b9c7d663fa0db: 609
BankBot/variety7/3fa3c06554aaaa243beb67bcbce6fdc0: 1415
BankBo

BankBot/variety4/0e17c4ba54e2fd4828ffd60510840fd1: 59
BankBot/variety4/173d496e0219ff78c57926d92971a880: 39
BankBot/variety4/4b1905902fcb03e9dc8f1755e5ac62e2: 52
BankBot/variety4/9420a82313688c76e8bf327a904df5f2: 52
BankBot/variety4/00d471c121078c8c72ddbd205015b40b: 47
BankBot/variety4/c429ff0736c751c2f4ff2d95c5c1fb4c: 47
BankBot/variety3/ab817db57e14bb9f9601473edb081ebc: 220
BankBot/variety3/262ce98bd2bcdcd4e256c235c734104f: 46
BankBot/variety3/477dc44c5078a23ec8ddbb93248284a6: 910
BankBot/variety3/eb4e730a37574507efa677339893e056: 910
BankBot/variety3/1628689265986904e143ae1cffd5d797: 46
BankBot/variety3/371be8d1513051ba83cf3b036efbb51c: 220
BankBot/variety3/d7d911bc406c97fb2204f38a547cf5f4: 910
BankBot/variety3/7d6f03323bed8f1bcdc2593be78d4835: 910
BankBot/variety3/4e2b5eb7fc21fc5e3303aa25efcf8caf: 910
BankBot/variety3/7ce37eb373f999dce924adc996f4fa20: 220
BankBot/variety3/31e0b319f135da9a02626ea5894261fd: 910
BankBot/variety3/a30f8cf8a2e5b3981baccadaefed1b7a: 910
BankBot/variety3/5

BankBot/variety8/d9ab0461e557b13a4ad2d89523879177: 1462
BankBot/variety8/c1d0adda208d6dc69c0604b233f5840a: 1462
BankBot/variety8/76921d8255ad8791ed2ef5b1f85ca268: 1462
BankBot/variety8/708a8a647e0203504a247ad9c8ffa6da: 1462
BankBot/variety8/ac7a09db46dc16c2af0be6f4f7e333e1: 939
BankBot/variety8/fcbc8de7ec411cd22c41dd9ef54aece2: 938
BankBot/variety8/ba59fda101be38880403337df729fb16: 938
BankBot/variety8/0c7882720a3e6888e9cc695a01127966: 1462
BankBot/variety8/5e1114bb090ccee2c29208eb2df82ea0: 1462
BankBot/variety8/ab69c5c0b420fee018a4b78cc7432797: 938
BankBot/variety8/5d9d78a122b263eab20db8684e8a4afb: 938
BankBot/variety8/1836d0549bb8aa08d0231f67499139b3: 1462
BankBot/variety8/ec919aa8599dc34f6ea395e3c363d40e: 938
BankBot/variety8/5f5d5560aa0a6a1015913696803be299: 1462
BankBot/variety8/a7df093ff16fbac0cad5fea19c506b61: 1462
BankBot/variety8/1857ac8d4da075b6f1413f788689df5f: 1462
BankBot/variety8/bec73a10547fabfb1e5fdb23de8d03ae: 1462
BankBot/variety8/1a01d0afe84c9577a6943cc9a60e0e08: 146

BankBot/variety2/26a4085917ad4fd4865ad12d8520445b: 358
BankBot/variety2/f83eceef159ed225c6c88409ca1880f3: 358
BankBot/variety2/b6c53e0c9281c5ca019adf3c7ad72630: 358
BankBot/variety2/fb834d01719cb78704c86a8aaed9f800: 358
BankBot/variety2/6bdd6a7ac049fc5c8d0f97d8e1e1682f: 358
BankBot/variety2/17ce17a6eeaa8bc7d362cad66d5b2aaf: 358
BankBot/variety2/fafd579a3086649a76bd310d83c5c3b8: 358
BankBot/variety2/93350854655cd78b24abfe2c10326cfc: 48
BankBot/variety2/794dba6ac1f2b45065513ea4e6cfe167: 358
BankBot/variety2/9a599ed71af893a72628ec981fde7d02: 358
BankBot/variety2/7e84f07f690b2132fb33cb2cd45974fa: 358
BankBot/variety2/93aae8420cbe553cbcc9a6ee0e609302: 358
BankBot/variety2/262508ecf92848aa7b5c246a5154cd13: 358
BankBot/variety2/d6a854b15ff63d7c50f249810d615ce4: 358
BankBot/variety2/8fc5a9397e84ded85fa9503a1ced1cdc: 48
BankBot/variety2/0f5f2701322a9ae2742810fc1db9d576: 358
BankBot/variety5/558d9d20cdc432172a8eb79ad823e5bc: 1090
BankBot/variety5/2b1eb4ce1509abb3b509ccf36084698d: 1090
BankBot/va

Looking at average class amounts, it seems a lot of malware is simply repackaged and rebranded as a different app, when the internals seem the same. For example, Stealer has either 120 or 122 classes, with no other values being present. The same goes for many other apps, such as BankBot having 48, 348, 938, or 1462 a lot of the time. Due to some malware having as low as 19 classes, and others having over 1000, class amount is not really indicative of malware when compared to benign apps.

In [2]:
import os
import networkx as nx
from tqdm import tqdm_notebook
from collections import defaultdict
import pandas as pd

config = {}
# config['path_benign'] = '/datasets/dsc180a-wi20-public/Malware/testing/benign'
# config['path_benign'] = 'test'
config['path_benign'] = 'data'
config['path_malware'] = '/datasets/dsc180a-wi20-public/Malware/testing/malware'

api_list = []
app_list = []
package_list = []
seen_api = set()
# api_counts = defaultdict(int)

# pandas method
# data = pd.DataFrame()

app_to_api = nx.Graph()
api_cooccur = nx.Graph()
api_same_invoke = nx.Graph()
api_same_package = nx.Graph()

for directory in tqdm_notebook(next(os.walk(config['path_benign']))[1]):
    app_list.append(directory)
    app_to_api.add_node(directory)
    
    # pandas method
#     app_data = []
    
    for subdir, dirs, files in os.walk(config['path_benign'] + '/' + directory):
        for i, file in enumerate(files):
            if i == 100:
                break
            filepath = subdir + os.sep + file

            with open(filepath, 'r') as fp:
                in_method = ''
                api_calls = set()
                
                for line in fp:
                    stripped = line.strip()
                    if stripped.startswith('.method'):
                        in_method = stripped[8:]
                    if stripped == '.end method':
                        in_method = ''
                        api_calls.clear()
                    if stripped[:6] == 'invoke':
                        invoke_method = stripped.split(' {')[0][7:].split('/')[0]
                        
                        splitted = line.split('}, ')
                        fns = splitted[1].split('->')
                        api_package = fns[0]
                        method = fns[1].split('(')[0]
                        
                        current_method_call = api_package + ',' + method
    
                        api_calls.add(current_method_call)
                        
                        # pandas method
#                         app_data.append([directory, invoke_method, '%s,%s,%s' % (directory, file, in_method), api_package, current_method_call])
    
                        if current_method_call not in seen_api:
                            api_list.append(current_method_call)
                            seen_api.add(current_method_call)
                    
                        # app to api generation
                        if current_method_call not in app_to_api:
                            app_to_api.add_node(current_method_call)
                        app_to_api.add_edge(directory, current_method_call)
                        
                        # api to api co-occurance generation
                        if in_method != '':
                            if current_method_call not in api_cooccur:
                                api_cooccur.add_node(current_method_call)
                            for api_call in api_calls:
                                if not api_cooccur.has_edge(current_method_call, api_call):
                                    api_cooccur.add_edge(current_method_call, api_call)
                                
                        # api to api invoke generation
                        # can go from invoke method <-> method call to method call <-> method call by doing A*(A^T)
                        if invoke_method not in api_same_invoke:
                            api_same_invoke.add_node(invoke_method)
                        if current_method_call not in api_same_invoke:
                            api_same_invoke.add_node(current_method_call)
                        if not api_same_invoke.has_edge(invoke_method, current_method_call):
                            api_same_invoke.add_edge(invoke_method, current_method_call)
                                
                        # api to api package generation
                        # can go from package <-> method call to method call <-> method call by doing A*(A^T)
                        if api_package not in api_same_package:
                            package_list.append(api_package)
                            api_same_package.add_node(api_package)
                        if current_method_call not in api_same_package:
                            api_same_package.add_node(current_method_call)
                        if not api_same_package.has_edge(api_package, current_method_call):
                            api_same_package.add_edge(api_package, current_method_call)
    # pandas method
#     data = data.append(app_data)
# data.columns = ['app', 'invoke', 'method', 'package', 'api']
# nx.draw(app_to_api)

In [33]:
# api-to-api same invoke
from scipy.sparse import csr_matrix
import scipy.sparse
import numpy as np
invoke_types = ['direct', 'static', 'virtual', 'super', 'interface']
matrix_P = nx.adjacency_matrix(api_same_invoke, nodelist=api_list + invoke_types)[-len(invoke_types):, :-len(invoke_types)]
# scipy.sparse.save_npz('matrix_C.npz', matrix_C)
# matrix_C_disk = np.memmap('matrix_C.npz', dtype='float32', mode='r')
# matrix_C_disk.shape
# matrix_C_disk_final = np.memmap('matrix_C_final.npz', dtype='float32', mode='w+', shape=(matrix_C_disk.shape[1], matrix_C_disk.shape[1]))
# matrix_C_disk_final[:] = matrix_C_disk.T.dot(matrix_C_disk)
matrix_P = matrix_P.transpose() @ matrix_P
# matrix_C = csr_matrix(np.matmul(matrix_C.T, matrix_C))
matrix_P

<17656x17656 sparse matrix of type '<class 'numpy.int64'>'
	with 103752202 stored elements in Compressed Sparse Column format>

In [34]:
# api-to-api same package
matrix_I = nx.adjacency_matrix(api_same_package, nodelist=api_list + package_list)[-len(package_list):, :-len(package_list)]
matrix_I = matrix_D.transpose() @ matrix_I
matrix_I

<17656x17656 sparse matrix of type '<class 'numpy.int64'>'
	with 239330 stored elements in Compressed Sparse Column format>

# pandas implementation for deriving matrices

In [ ]:
# pandas app-api
from scipy.sparse import csr_matrix
matrix_A = csr_matrix(pd.pivot_table(data, index='app', columns='api', aggfunc='count').fillna(0).clip(0, 1).values)
matrix_A

In [12]:
# pandas api-api same package
from scipy.sparse import csr_matrix
import numpy as np
part_package = csr_matrix(pd.pivot_table(data, index='api', columns='package', aggfunc='count').fillna(0).clip(0, 1).values)
matrix_package = part_package @ part_package.transpose()
matrix_package

<17656x17656 sparse matrix of type '<class 'numpy.float64'>'
	with 239330 stored elements in Compressed Sparse Row format>

# Analysis on benign apps

In [8]:
# O(n + k) top kth frequent appearing elements, sorry heap but O(nlogk) is just too slow for leetcode

freq_map = defaultdict(list)
max_degree = 0
for node in app_to_api.degree:
    if max_degree < node[1]:
        max_degree = node[1]
    freq_map[node[1]].append(node[0])

highest_k = 100
should_break = False
for i in range(len(app_list), -1, -1):
    if i in freq_map:
        for api in freq_map[i]:
            print('%s: %i' % (api, i))
            highest_k -= 1
            if highest_k == 0:
                should_break = True
                break
    if should_break:
        break

Ljava/lang/Object;,<init>: 98
Landroid/app/Activity;,<init>: 98
Ljava/lang/StringBuilder;,<init>: 97
Ljava/lang/StringBuilder;,append: 97
Ljava/lang/StringBuilder;,toString: 97
Ljava/lang/Integer;,valueOf: 97
Ljava/util/ArrayList;,<init>: 97
Ljava/util/List;,size: 97
Ljava/util/Iterator;,hasNext: 97
Ljava/util/Iterator;,next: 97
Landroid/content/Context;,getPackageName: 97
Ljava/util/ArrayList;,add: 97
Ljava/lang/System;,arraycopy: 97
Ljava/lang/RuntimeException;,<init>: 97
Landroid/content/Intent;,<init>: 97
Ljava/lang/Object;,getClass: 97
Landroid/net/Uri;,parse: 97
Ljava/lang/Boolean;,valueOf: 97
Ljava/util/Locale;,getDefault: 97
Ljava/util/List;,iterator: 97
Landroid/app/Activity;,onCreate: 97
Landroid/content/Context;,getSystemService: 96
Landroid/view/WindowManager;,getDefaultDisplay: 96
Ljava/lang/Math;,min: 96
Ljava/lang/IllegalArgumentException;,<init>: 96
Landroid/graphics/Rect;,<init>: 96
Landroid/content/res/Resources;,getDisplayMetrics: 96
Landroid/util/Log;,w: 96
Ljava/la

The app with the highest amount of API calls in my sample set is a cryptocurrency wallet, with a total of 63992 unique API calls. The API with the highest usages are core Java utils, used among all 50 apps. These include things such as initialization of Object (from super() calls due to inheritance), StringBuilders, Java Collection objects, and IOStreams. However, these are boring, so above shown are only APIs that are called at most (# of apps - 1) times. In the same vein, the code below shows the least-used 100 APIs that are used by at least 2 apps. Many of these APIs have to do with ads, or analytics.

In [10]:
highest_k = 100
should_break = False
for i in range(2, len(app_list), 1):
    if i in freq_map:
        for api in freq_map[i]:
            print('%s: %i' % (api, i))
            highest_k -= 1
            if highest_k == 0:
                should_break = True
                break
    if should_break:
        break

[Lcom/google/ads/b;,clone: 2
[Lcom/google/ads/c;,clone: 2
Lcom/google/ads/mediation/d;,b: 2
Lcom/google/ads/mediation/d;,c: 2
Lcom/google/ads/mediation/d;,d: 2
Lcom/google/android/gms/ads/mediation/MediationBannerListener;,a: 2
Lcom/google/android/gms/ads/mediation/MediationBannerListener;,b: 2
Lcom/google/android/gms/ads/mediation/MediationBannerListener;,d: 2
Lcom/google/android/gms/ads/mediation/MediationBannerListener;,c: 2
Lcom/google/android/gms/ads/mediation/MediationBannerListener;,e: 2
Lcom/google/android/gms/ads/mediation/MediationInterstitialListener;,b: 2
Lcom/google/ads/mediation/c;,<init>: 2
Lcom/google/ads/mediation/d;,<init>: 2
Lcom/google/android/gms/ads/mediation/MediationAdRequest;,a: 2
Lcom/google/android/gms/ads/mediation/MediationAdRequest;,b: 2
Lcom/google/android/gms/ads/mediation/MediationAdRequest;,c: 2
Lcom/google/android/gms/ads/mediation/MediationAdRequest;,d: 2
Lcom/google/android/gms/ads/internal/client/zzn;,a: 2
Lcom/google/android/gms/ads/internal/util/

# Malware
To do malware analysis here, one variant of each malware type will be parsed.

In [3]:
import os
import networkx as nx
from tqdm import tqdm_notebook
from collections import defaultdict
import pandas as pd

config = {}
# config['path_benign'] = '/datasets/dsc180a-wi20-public/Malware/testing/benign'
# config['path_benign'] = 'test'
config['path_benign'] = 'data'
config['path_malware'] = '/datasets/dsc180a-wi20-public/Malware/amd_data_smali'
# api_counts = defaultdict(int)

# pandas method
# data = pd.DataFrame()

# app_to_api = nx.Graph()
# api_cooccur = nx.Graph()
# api_same_invoke = nx.Graph()
# api_same_package = nx.Graph()

for directory in tqdm_notebook(next(os.walk(config['path_malware']))[1]):
    app_list.append(directory)
    app_to_api.add_node(directory)
    
    for subdir, dirs, files in os.walk(config['path_malware'] + '/' + directory + '/' + os.listdir(config['path_malware'] + '/' + directory)[0]):
        for i, file in enumerate(files):
            if i == 100:
                break
            filepath = subdir + os.sep + file

            with open(filepath, 'r') as fp:
                in_method = ''
                api_calls = set()

                for line in fp:
                    stripped = line.strip()
                    if stripped.startswith('.method'):
                        in_method = stripped[8:]
                    if stripped == '.end method':
                        in_method = ''
                        api_calls.clear()
                    if stripped[:6] == 'invoke':
                        invoke_method = stripped.split(' {')[0][7:].split('/')[0]

                        splitted = line.split('}, ')
                        fns = splitted[1].split('->')
                        api_package = fns[0]
                        method = fns[1].split('(')[0]

                        current_method_call = api_package + ',' + method

                        api_calls.add(current_method_call)

                        # pandas method
#                         app_data.append([directory, invoke_method, '%s,%s,%s' % (directory, file, in_method), api_package, current_method_call])

                        if current_method_call not in seen_api:
                            api_list.append(current_method_call)
                            seen_api.add(current_method_call)

                        # app to api generation
                        if current_method_call not in app_to_api:
                            app_to_api.add_node(current_method_call)
                        app_to_api.add_edge(directory, current_method_call)

                        # api to api co-occurance generation
                        if in_method != '':
                            if current_method_call not in api_cooccur:
                                api_cooccur.add_node(current_method_call)
                            for api_call in api_calls:
                                if not api_cooccur.has_edge(current_method_call, api_call):
                                    api_cooccur.add_edge(current_method_call, api_call)

                        # api to api invoke generation
                        # can go from invoke method <-> method call to method call <-> method call by doing A*(A^T)
                        if invoke_method not in api_same_invoke:
                            api_same_invoke.add_node(invoke_method)
                        if current_method_call not in api_same_invoke:
                            api_same_invoke.add_node(current_method_call)
                        if not api_same_invoke.has_edge(invoke_method, current_method_call):
                            api_same_invoke.add_edge(invoke_method, current_method_call)

                        # api to api package generation
                        # can go from package <-> method call to method call <-> method call by doing A*(A^T)
                        if api_package not in api_same_package:
                            package_list.append(api_package)
                            api_same_package.add_node(api_package)
                        if current_method_call not in api_same_package:
                            api_same_package.add_node(current_method_call)
                        if not api_same_package.has_edge(api_package, current_method_call):
                            api_same_package.add_edge(api_package, current_method_call)
    # pandas method
#     data = data.append(app_data)
# data.columns = ['app', 'invoke', 'method', 'package', 'api']
# nx.draw(app_to_api)

In [3]:
# O(n + k) top kth frequent appearing elements, sorry heap but O(nlogk) is just too slow for leetcode

freq_map = defaultdict(list)
max_degree = 0
for node in app_to_api.degree:
    if max_degree < node[1]:
        max_degree = node[1]
    freq_map[node[1]].append(node[0])

highest_k = 100
should_break = False
for i in range(len(app_list), -1, -1):
    if i in freq_map:
        for api in freq_map[i]:
            print('%s: %i' % (api, i))
            highest_k -= 1
            if highest_k == 0:
                should_break = True
                break
    if should_break:
        break

Ljava/lang/Object;,<init>: 33
Ljava/lang/StringBuilder;,<init>: 33
Ljava/lang/StringBuilder;,append: 33
Ljava/lang/StringBuilder;,toString: 33
Ljava/lang/String;,equals: 33
Landroid/os/Handler;,<init>: 33
Landroid/app/Activity;,<init>: 33
Landroid/app/Activity;,onCreate: 33
Ljava/util/ArrayList;,<init>: 32
Ljava/lang/String;,length: 32
Ljava/lang/String;,substring: 32
Ljava/lang/Integer;,valueOf: 32
Ljava/util/Iterator;,hasNext: 32
Ljava/util/Iterator;,next: 32
Landroid/content/Intent;,<init>: 32
Ljava/lang/System;,currentTimeMillis: 32
Ljava/util/List;,get: 31
Landroid/content/Context;,getSystemService: 31
Ljava/lang/Integer;,intValue: 31
Landroid/content/BroadcastReceiver;,<init>: 31
Landroid/content/Intent;,putExtra: 31
Ljava/lang/Integer;,parseInt: 31
Landroid/net/Uri;,parse: 31
Landroid/util/Log;,d: 31
Landroid/util/Log;,e: 30
Ljava/util/List;,add: 30
Ljava/lang/reflect/Method;,invoke: 30
Ljava/lang/Thread;,<init>: 30
Ljava/lang/String;,startsWith: 30
Ljava/lang/Object;,getClass: 

In [4]:
highest_k = 100
should_break = False
for i in range(2, len(app_list), 1):
    if i in freq_map:
        for api in freq_map[i]:
            print('%s: %i' % (api, i))
            highest_k -= 1
            if highest_k == 0:
                should_break = True
                break
    if should_break:
        break

Landroid/graphics/drawable/Drawable$ConstantState;,getChangingConfigurations: 2
Landroid/support/v4/content/res/ResourcesCompat;,getDrawable: 2
Landroid/animation/AnimatorSet;,getChildAnimations: 2
Landroid/animation/ObjectAnimator;,getPropertyName: 2
Landroid/animation/ArgbEvaluator;,<init>: 2
Landroid/animation/ObjectAnimator;,setEvaluator: 2
Landroid/support/v4/graphics/drawable/DrawableCompat;,getAlpha: 2
Landroid/support/v4/graphics/drawable/DrawableCompat;,inflate: 2
Landroid/animation/AnimatorInflater;,loadAnimator: 2
Landroid/graphics/drawable/AnimatedVectorDrawable;,isRunning: 2
Landroid/graphics/drawable/AnimatedVectorDrawable;,start: 2
Landroid/graphics/drawable/AnimatedVectorDrawable;,stop: 2
Landroid/graphics/drawable/Drawable;,onLevelChange: 2
Landroid/graphics/PathMeasure;,setPath: 2
Landroid/graphics/PathMeasure;,getSegment: 2
Landroid/support/v4/graphics/ColorUtils;,getTempDouble3Array: 2
Landroid/support/v4/graphics/ColorUtils;,LABToXYZ: 2
Landroid/support/v4/graphics

Looking at malware used APIs, some APIs immediately jump to the forefront. There are uses of os-level APIs from android.os, as well as usage of typed arrays, probably for some sort of buffer overflow attack. Commonly used APIs include setting shared preferences, hex strings, file operations, and intent. Less commonly-used APIs for malware do not include ad-related APIs like benign apps. Most of these APIs seem to be for backwards graphical support.

In [4]:
# testapp-api @ api-api @ api-testapp -> score for models
# app-api
matrix_A = nx.adjacency_matrix(app_to_api, api_list + app_list)[-len(app_list):, :-len(app_list)]

# api-api co-occur
matrix_B = nx.adjacency_matrix(api_cooccur, api_list)

matrix_A, matrix_B

(<132x1163654 sparse matrix of type '<class 'numpy.int64'>'
 	with 2756767 stored elements in Compressed Sparse Row format>,
 <1163654x1163654 sparse matrix of type '<class 'numpy.int64'>'
 	with 37935736 stored elements in Compressed Sparse Row format>)

In [5]:
from sklearn.linear_model import LogisticRegression
import scipy.sparse
app_to_app = matrix_A @ matrix_B @ matrix_A.transpose()
# scipy.sparse.save_npz('matrix_ABAT.npz', app_to_app)
# app_to_app = scipy.sparse.load_npz('matrix_ABAT.npz')
y_list = [1] * len(app_list[:-33]) + [0] * len(app_list[-33:])

In [6]:
logistic = LogisticRegression()
logistic.fit(app_to_app[5:-5], y_list[5:-5])
logistic.score(app_to_app[:5], y_list[:5]), logistic.score(app_to_app[-5:], y_list[-5:])

/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


(0.8, 1.0)

With a simple logistic regression, 94 benign samples, and 28 malware samples, there is an accuracy of 80% on scoring benign apps as benign, and all 5 malware samples were identified as malware. The commuting matrix used was ABA<sup>T</sup>. This commuting matrix was used because B is the co-occur matrix, so apps that use the same APIs in one method are probably closely related in how they function.

In [7]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(app_to_app[5:-5], y_list[5:-5])
rfc.score(app_to_app[:5], y_list[:5]), logistic.score(app_to_app[-5:], y_list[-5:])

/opt/conda/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


(1.0, 1.0)

In [10]:
from xgboost import XGBClassifier
xgb = XGBClassifier(n_jobs=8, booster='dart', tree_method='exact', n_estimators=200)
xgb.fit(app_to_app[5:-5], y_list[5:-5])
xgb.score(app_to_app[:5], y_list[:5]), logistic.score(app_to_app[-5:], y_list[-5:])

(1.0, 1.0)

Using a random forest and xgboost library both led to perfect accuracy scores on benign and malware. However, this is a limited test set and such results aren't indicative of all different varieties of malware and benign apps. Most apps collected here were games, so productivity apps and other categories are left out and may be categorized as malware accidentally.